lib/fibonacci.rb

```diff
@@ -5,8 +5,23 @@
 # ....
 # e.g. 6th fibonacci number is 8
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) because the number of iterations is dependent on the size of the number
+# Space complexity: O(1) because there will only ever be 4 int variables created, regardless of the size of the number
 def fibonacci(n)
-  raise NotImplementedError
+  raise ArgumentError, "n must be an integer greater than 0" unless n && n >= 0
+  current = 0
+  one_before = 0
+  two_before = 0
+  i = 0
+  while i <= n
+    if i == 1
+      current = 1
+    else
+      two_before = one_before
+      one_before = current
+      current = one_before + two_before
```

Comment:
Understanding logic like this will help you tremendously when you start working with linked lists. Nice work on this.